### PR TITLE
Fix several issues with the GPU support

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR has fixes for multiple issues with the GPU support. All of these are
motivated by the desire to be able to use distributed arrays to allocate arrays
on GPUs. Sadly, this PR does not fully enable that. It only fixes some smaller
issues that I bumped into on the way.

The problems that this PR fixes:

- Removes the limitation on the outliner that caused it to look at only user
  modules.
- Adds `chpl-gpu-gen-includes.h` header to support runtime calls added at
  codegen time in GPU kernels.
- Adjust chplenv scripts to be able to put that file into
  `runtime/include/gpu/cuda`
- Adds a new environment variable `CHPL_GPU_CODEGEN`. Currently this could
   be only `cuda` or `none`. If you set `CHPL_LOCALE_MODEL=gpu`, then we
   pick `cuda` automatically without any check.
- Handle `PRIM_GET_MEMBER` in loop bodies properly.
- Set a GPU sublocale's maximum task parallelism to be 1. Previously, it was
  defaulting to 0.
- Fix a bug with 0-sized allocations/frees in the GPU allocators. CUDA Driver
  API functions we use require non-zero size, and they'll return an error
  otherwise. We now check for the non-zero size before calling those functions.
- Adds new tests.
- Adds a new prediff filter to suppress a ptxas warning. This is almost certainly
  not the right thing to do. See https://github.com/Cray/chapel-private/issues/2758

A nice side-effect of this PR is that now we can run simple promotions on GPU.
I was able to run `A = 1` on GPU, but not `A = B * alpha + C`. (status of 
promotion on GPU is in https://github.com/Cray/chapel-private/issues/2695)

Test:
- [x] gpu/native

### CODE REVIEW

**Comment by e-kayrakli** [comment]
@jhh67 / @ronawho / @gbtitus / anyone interested in the runtime: Currently https://github.com/chapel-lang/chapel/pull/18762/files#diff-b02d3937e807ce7d443313812aad74ae5cb6da7482520e814fe75ee3d860a2f8 is the file that causes CI issues, because we don't have a `gpu/none` version of it. Should we:

1. move this file into the root `runtime/include` and revert all the Makefile/chplenv changes
2. add `gpu/none` counterpart of it that's basically empty
3. add `#if HAS_GPU_LOCALE` around its include in `stdchpl.h` ?

In the near term I would like to be able to include that file from somewhere more GPU specific. So, this'll hopefully be moot. My preference would be 3>2>1, but that's probably the exact opposite order from the runtime structure and maintainability standpoint.

Thoughts?

**Comment by ronawho** [comment]
> 3. add #if HAS_GPU_LOCALE around its include in stdchpl.h ?

Discussed offline, but I'd go with this for now. Seems simplest to make progress without over-engineering. Even if we leave it in the long term it does not offend me, and if we need something more complicated in the future (for other vendors, or whatever) it'll be easy to change still.

**Review by daviditen — APPROVED** [review]

File: util/chplenv/chpl_compiler.py [hunk 1]
```diff
@@ -197,6 +198,12 @@ def get(flag='host'):
                 compiler_val = 'clang'
             else:
                 compiler_val = 'gnu'
+        elif locale_model_val == 'gpu':
+            if find_executable('clang'):
+                compiler_val = 'clang'
+            else:
+                error("clang not found. The 'gpu' locale model is supported "
+                      "with clang only.")
```
Comment (mppf) [inline util/chplenv/chpl_compiler.py:206]: This has the effect of requiring that `chpl` be built with clang (i.e. the compiler source .cpp files are compiled with clang). That seem surprising to me. Is that expected? Anyway this kind of check needs to go in to validate_compiler.

Comment (e-kayrakli) [reply util/chplenv/chpl_compiler.py:206]: I wasn't thinking about requiring that here, and intended this as something that can give an early error message if you are using GPU locale model without clang.

Comment (mppf) [reply util/chplenv/chpl_compiler.py:206]: Do you mean "using the GPU locale model without CHPL_TARGET_COMPILER=llvm" ?

Comment (e-kayrakli) [reply util/chplenv/chpl_compiler.py:206]: Don't we need `clang`? Gotta say that the difference between the two in this context is not crystal clear to me.

**Pull request merged**